enhancements/cvo/upgrade_preflight_checks.md

```diff
@@ -0,0 +1,107 @@
+
+## Motivation
+
+CVO managed operators should be able to run preflight checks before the actual upgrade.
+It would help operators to find out if they can move to the specific version safely before doing an actual upgrade.
+Also administrators are empowered to run these preflight checks without doing an upgrade.
+
+
+### Goals
+
+* Administrators will be able to run preflight checks without running the upgrade.
+* Give Operators (CVO managed) ability to add checks to check if they can safely upgrade to a given release.
+* Preflight checks would be idempotent in nature.
+
+### Non-Goals
+
+* This enhancement will not add code required for preflight checks for individual operators.
+* Operators will have to figure out the actual code they would like to run as part of preflight checks.
+* The preflight checks should not be used as a migration mechanism.
+
+## Proposal
+
+Operators need to provide a different bin path for the preflight checks i.e. /bin/preflight.
+The binpath need to exist inside of the operator image.
+Cluster version operator (CVO) will be running a single job per operator for the preflight checks.
+The binpath /bin/preflight must be executable and it can be symbolic link to the operator binary or a script or It can be non existent.
+CVO will not fail if the path does not exist in an operator image but it would assume that preflight checks are not implemented for the respective operator.
+Also there would be a possible way to skip the preflight checks if someone wants to skip it explicitly.
+
+### User Stories
+
+#### Story 1
+
+Some customers would prefer to disable automatic credential provisioning by Cloud credential operator (CCO) and instead manage it themselves.
+However situations can arise during an upgrade where credentials have changed and the upgrade will be block and go into a failed state, with little visibility as to why or how it can be fixed.
+With the help of preflight checks CCO can check if they can safely upgrade to a given release before doing the actual upgrade.
+Communicate that an upgrade is not possible as permissions have changed and point administrator with right pointers to fix the issue.
+This would allow users to avoid broken upgrades when credentials have changed and provides them right information to move forward.
+
+### Implementation Details
+
+CVO will run the preflight check from the new operator image to which the current operator image is trying to upgrade.
+Because the current version of operator does not have knowledge of what new code changes the new version bringing with it.
+So its the new version's responsibility to figure out if the current image can update safely to the new image.
+The preflight jobs would run in the same namespace as operators as it would need same RBAC as the operators.
```

Comment:
Hrm... this would make [running preflights for all referenced images][1] hard, and would also block [using read-only RBAC to keep preflights from making changes][2].  Can we run the jobs in the `openshift-cluster-version` namespace (or a new `openshift-update-preflight` namespace, or some such?), and have this enhancement define the RBAC policy that gets applied?  Then the RBAC decisions would be baked into the source release image, but I'm more confident in our ability to figure out that RBAC ahead of time than I am about our ability to figure out the operator-specific logic ahead of time.

[1]: https://github.com/openshift/enhancements/pull/363#discussion_r436984815
[2]: https://github.com/openshift/enhancements/pull/363#discussion_r436968647

Reply:
Could we drop the "same RBAC as the operator" in favor of global read-only RBAC, but still run in the operator's namespace?

Reply:
What does running in the operator gain you (for images where you can figure it out)?  Besides the two issues I raise above (image -> operator identification and read-only RBAC), this would also break [CVO volume sharing][1], right?

[1]: https://github.com/openshift/enhancements/pull/363#discussion_r437002776

Reply:
Yeah that's a good case for the CVO ns. I suppose we can access any data we need via RBAC regardless and there shouldn't be anything needing to mount in from the operator namespace regardless. Your approach would also work for net-new operators coming in that don't already have a namespace in the cluster.

Reply:
I feel like this case of a net new operator needs to be included in the goals section or somewhere else.